transport/src/test/java/io/netty/channel/DefaultChannelPipelineTest.java

```diff
@@ -608,6 +614,416 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+            String handlerName = "foo";
+            final Exception exception = new RuntimeException();
+            ChannelPipeline pipeline = new LocalChannel().pipeline();
+            pipeline.addLast(handlerName, new ChannelHandlerAdapter() {
+                @Override
+                public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+                    throw exception;
+                }
+            });
+            pipeline.addLast(group1, new CheckExceptionHandler(exception, promise));
+            group.register(pipeline.channel()).syncUninterruptibly();
+            pipeline.remove(handlerName);
+            promise.syncUninterruptibly();
+        } finally {
+            group1.shutdownGracefully();
+        }
+    }
+
+    @Test(timeout = 3000)
+    public void testHandlerAddedThrowsAndRemovedThrowsException() {
+        final EventExecutorGroup group1 = new DefaultEventExecutorGroup(1);
+        try {
+            final Promise<Void> promise = group1.next().newPromise();
+            final Exception exceptionAdded = new RuntimeException();
+            final Exception exceptionRemoved = new RuntimeException();
+            String handlerName = "foo";
+            ChannelPipeline pipeline = new LocalChannel().pipeline();
+            pipeline.addLast(handlerName, new ChannelHandlerAdapter() {
+                @Override
+                public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+                    throw exceptionAdded;
+                }
+
+                @Override
+                public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+                    throw exceptionRemoved;
+                }
+            });
+            pipeline.addLast(group1, new CheckExceptionHandler(exceptionAdded, promise));
+            group.register(pipeline.channel()).syncUninterruptibly();
+            assertNull(pipeline.context(handlerName));
+            promise.syncUninterruptibly();
+        } finally {
+            group1.shutdownGracefully();
+        }
+    }
+
+    @Test(timeout = 3000)
+    public void testHandlerAddBlocksUntilHandlerAddedCalled() {
+        final EventExecutorGroup group1 = new DefaultEventExecutorGroup(1);
+        try {
+            final Promise<Void> promise = group1.next().newPromise();
+            ChannelPipeline pipeline = new LocalChannel().pipeline();
+            group.register(pipeline.channel()).syncUninterruptibly();
+
+            pipeline.addLast(new ChannelHandlerAdapter() {
+                @Override
+                public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+                    final AtomicBoolean handlerAddedCalled = new AtomicBoolean();
+                    ctx.pipeline().addLast(group1, new ChannelHandlerAdapter() {
+                        @Override
+                        public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+                            Thread.sleep(100);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)